crates/sparrow-compiler/src/data_context.rs

```diff
@@ -303,6 +303,52 @@ pub struct GroupInfo {
     key_type: DataType,
 }
 
+/// Concurrent file sets allows us to update the file sets from the
+/// python library and see updates reflected within the data context.
+#[derive(Debug, Default, Clone)]
+pub struct ConcurrentFileSets {
+    file_sets: Arc<Mutex<Vec<compute_table::FileSet>>>,
+}
+
+impl ConcurrentFileSets {
+    pub fn new(file_sets: Vec<compute_table::FileSet>) -> Self {
+        Self {
+            file_sets: Arc::new(Mutex::new(file_sets)),
+        }
+    }
+
+    pub fn push(&mut self, file_set: compute_table::FileSet) {
```

Comment:
I think having this be `&mut self` is misleading -- it makes us think the ownership is what let's us modify. Generally, I'd suggest:

1. having `file_sets` be `Mutex<...>` rather than `Arc<Mutex<...>>`
2. having these take `&self` to indicate "interior mutability" (eg., you can mutate this with only a reference because ti has a mutex inside it)
3. having users choose to have a `Arc<ConcurrentFileSets>` for sharing -- this makes it clear to them that it has an arc and is cheap to clone.

But, I'm not 100% sure that's the best pattern -- just what I've tended towards. We can discuss further.

Reply:
Ah, yep. That makes sense, made that change.